### PR TITLE
[tf.data] graduate experimental `get_single_element` API to tf.data

### PR DESCRIPTION
**Pull request by kvignesh1420**
UPDATED: This PR graduates the `tf.data.experimental.get_single_element` API into `tf.data.get_single_element` by making the following changes:

 - [x] Adds the deprecation decorator for the experimental API.
 - [x] Adds the necessary `group_by_window` method to `DatasetV2` class.
 - [x] Updates example in documentation with new API.
 - [x] Regenerate golden API's.
 - [x] Moved and updated the `get_single_element_test` test target from experimental/kernel_tests to kernel_tests

TEST LOG
```
INFO: Build completed successfully, 343 total actions
//tensorflow/python/data/kernel_tests:get_single_element_test            PASSED in 3.7s

```
cc: @jsimsa

### CODE REVIEW

**Comment by jsimsa** [comment]
Thanks @kvignesh1420. I think we will have to decouple this PR into multiple separate steps because we cannot really remove the `tf.data.experimental.get_single_element` method without breaking a lot of users, which should be avoided.

My suggestion would be for this PR do the following:
1) Add "data.get_single_element" into the `tf_export` decorator of the current implementation and deprecate the experimental endpoint ([example](https://github.com/tensorflow/tensorflow/blob/v2.4.1/tensorflow/python/data/ops/optional_ops.py#L35-L36))
2) Update the documentation.

In particular, you should not try to remove the "tf.data.experimental.get_single_element" endpoint or move the current source / test code location. Because is strictly speaking not necessary to graduate the API to non-experimental and could cause churn and breakages.

**Comment by kvignesh1420** [comment]
@jsimsa thanks for the info. I have made the changes to deprecate the experimental API and expose it in `tf.data`.
Also, please let me know your opinion on the documentation. I tried to keep the code sample consistent between examples so that it's easy to understand. Thanks!

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 1]
```diff
@@ -20,31 +20,94 @@
...
+  dataset = (tf.data.Dataset.from_tensor_slices(raw_features)
              .map(preprocessing_fn, num_parallel_calls=BATCH_SIZE)
              .batch(BATCH_SIZE))
 
-  image_batch, label_batch = tf.data.experimental.get_single_element(dataset)
+  processed_features = tf.data.get_single_element(dataset)
+  ```
+
+  In the above example, the `raw_features` tensor of length=BATCH_SIZE
+  were converted to a `tf.data.Dataset`. Next, all the raw_features were
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:55]: s/were/was

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 2]
```diff
@@ -20,31 +20,94 @@
...
+  mapped using the `preprocessing_fn` and the processed features were
+  grouped into a single batch. The final `dataset` contains only one element
+  which is a batch of all the processed features.
+
+  NOTE: The `dataset` should contain only one element. Preferrably, a batch
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:60]: Remove the sentence starting with "Preferably, ..."

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 3]
```diff
@@ -20,31 +20,94 @@
...
+  For example:
+
+  ```python
+  def serving_input_fn():
+    # The function transforms the raw features of the data that is fed to
+    # the model. Generally, an input_fn that expects data to be fed to the
+    # model at serving time is used to get the raw_features and the tensor
+    # placeholders.
+    #
+    # Reference: https://www.tensorflow.org/tfx/tutorials/transform/census
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:87]: remove this part

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 4]
```diff
@@ -20,31 +20,94 @@
...
+
+    input_fn = ... # estimator based data receiver function
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:89]: would it make sense to use `tf.estimator.export.build_raw_serving_input_receiver_fn()` here?

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 5]
```diff
@@ -20,31 +20,94 @@
...
+    dataset = (tf.data.Dataset.from_tensor_slices(raw_features)
+              .map(preprocessing_fn, num_parallel_calls=BATCH_SIZE)
+              .batch(BATCH_SIZE))
+
+    processed_features = tf.data.get_single_element(dataset)
+
+    return tf.estimator.export.ServingInputReceiver(
+        processed_features, serving_input_receiver.receiver_tensors
+        # The `serving_input_receiver.receiver_tensors` are placeholder
+        # for the `processed_features`
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:106]: I don't think that this comment adds much value because I can't make sense of it.

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 6]
```diff
@@ -20,31 +20,94 @@
...
+  and inferecing using [estimator](https://www.tensorflow.org/guide/estimator)
+  models.
+
+  In order to export the saved estimator model, you need to generally define a
+  `serving_input_fn` which would require the features to be processed by the
+  model while inferencing.
+
+  For example:
+
+  ```python
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:80]: On a second thought I wonder if it would make sense to illustrate how to use the API with Keras ([example](https://sayak.dev/tf.keras/preprocessing/2020/04/13/embedding-image-preprocessing-functions.html#Step-3:-SavedModel-plunge)) instead of Estimator seeing as Estimator is a deprecated API.

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 1 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py:55]: addressed in the commit.

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 2 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py]: addressed in the commit.

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 4 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py]: addressed in the commit.

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 5 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py]: addressed in the commit.

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 6 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py:157]: addressed in the commit.

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 3 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py]: addressed in the commit.

**Comment by kvignesh1420** [comment]
@jsimsa thanks for the references, I was able to document a fully working `tf.keras` based example. I kept the estimator based example as well just in case someone wants to use it. However, the lack of proper documentation around `tf.estimator.export.ServingInputReceiver` made it a bit difficult to document the details. Please check and let me know.

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 7]
```diff
@@ -20,31 +20,175 @@
...
+      return tf.argmax(probabilities, axis=-1)
+
+  preprocessing_model = PreprocessingModel(model)
+  fmnist_save_path = "fmnist"
+
+  tf.saved_model.save(preprocessing_model, fmnist_save_path,
+                signatures={'serving_default': preprocessing_model.serving_fn})
+  ```
+
+  The saved model can now be served using `tensorflow-serving`.
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:137]: remove the part about tensorflow serving and estimator

Comment (jsimsa) [reply tensorflow/python/data/experimental/ops/get_single_element.py:137]: actually, keep the estimator but remove the tensorflow serving part

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 8]
```diff
@@ -20,31 +20,175 @@
...
+  model = make_model()
+  train_data = make_dataset(x_train, y_train)
+  test_data = make_dataset(x_test, y_test)
+  model.fit(train_data, validation_data=test_data, epochs=10)
+
+  # There is a need to pre-process the data before inference as the model
+  # will throw exceptions if the input is not of the desired shape. Thus, to
+  # serialize and save the pre-processing functionality along with the model,
+  # a `PreprocessingModel` is defined with a custom serving function called
+  # `serving_fn`.
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:112]: Simplify this as follows:

```
model = ...

class PreprocessingModel(tf.keras.Model):
  def __init__(self, model):
    super().__init__(self)
    self.model = model

  @tf.function(input_signature=[...])
  def serving_fn(self, data):
    ds = tf.data.Dataset.from_tensor_slices(data)
    ds = ds.map(preprocessing_fn, num_parallel_calls=BATCH_SIZE)
    ds = ds.batch(batch_size=BATCH_SIZE)
    return tf.argmax(self.model(tf.data.experimental.get_single_element(ds)), axis=-1)

preprocessing_model = PreprocessingModel(model)
tf.saved_model.save(preprocessing_model, your_exported_model_path,
                signatures={'serving_default': preprocessing_model.serving_fn})
```

Doing this will provide consistency between the tf.keras and tf.estimator example.

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 9]
```diff
@@ -20,31 +20,175 @@
...
+  ```console
+  $ apt-get install tensorflow-model-server
+  $ nohup tensorflow_model_server \
+    --rest_api_port=8501 \
+    --model_name=fashion_model \
+    --model_base_path="fmnist" >server.log 2>&1
+  ```
+  For additional details, refer: https://www.tensorflow.org/tfx/serving/serving_basic.
+
+  Estimators:
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:148]: tf.estimator

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 10]
```diff
@@ -20,31 +20,175 @@
...
+
+  In the case of estimators, you need to generally define a `serving_input_fn`
+  which would require the features to be processed by the model while
+  inferencing.
+
+  For example:
+
+  ```python
+  def serving_input_fn():
+    # The function transforms the raw features of the data that is fed to
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:159]: remove comments that are not related to tf.data (such as this one) from this function ... the documentation of `get_single_element` should focus on tf.data specific logic and there is no need to explain tf.estimator specific logic

**Review by jsimsa — COMMENTED** [review]
Thank you. Did another pass and left a couple of comments. As is the documentation is a bit too verbose so most of my comments aim to make it more succinct.

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 7 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py]: addressed in the commit.

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 8 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py]: addressed in the commit.

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 10 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py:149]: yes, makes sense.

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 9 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py]: addressed in the commit.

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 11]
```diff
@@ -20,31 +20,118 @@
...
+    serving_input_receiver = input_fn()
+    raw_features = serving_input_receiver.features
+
+    def preprocessing_fn(raw_feature):
+      # ... the raw_feature is preprocessed as per the use-case
+      return feature
+
+    dataset = (tf.data.Dataset.from_tensor_slices(raw_features)
+              .map(preprocessing_fn, num_parallel_calls=BATCH_SIZE)
+              .batch(BATCH_SIZE))
```
Comment (kvignesh1420) [inline tensorflow/python/data/experimental/ops/get_single_element.py:126]: @jsimsa should we add some info on what the `BATCH_SIZE` can be? It should ideally be equal to the number of elements in `raw_features`. WDYT?

Comment (jsimsa) [reply tensorflow/python/data/experimental/ops/get_single_element.py:126]: Sounds good. Adding a comment that the code assume that the leading dimension of `raw_features` is `BATCH_SIZE` should be sufficient.

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 12]
```diff
@@ -20,31 +20,118 @@
...
+  NOTE: The `dataset` should contain only one element.
+
+  Now, instead of creating an iterator for the `dataset` and retrieving the
+  batch of features, the `tf.data.get_single_element()` function is used
+  to skip the iterator creation process and directly output the batch of
+  features.
+
+  This can be particularly useful when your tensor transformations are
+  expressed as `tf.data.Dataset` operations, and you want to use those
+  transformations while serving your model. Especially, when it comes to
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:69]: remove the "Especially, ..." sentence.

**Review by jsimsa — COMMENTED** [review]
Almost there, a couple of last small comments.

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 12 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py]: done!

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 11 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py:125]: done!

**Comment by kvignesh1420** [comment]
@jsimsa done!

File: tensorflow/python/data/experimental/ops/get_single_element.py (hunk 8 above)
Comment (jsimsa) [reply tensorflow/python/data/experimental/ops/get_single_element.py]: Please update the body of `serving_fn` as per my suggestion.

Comment (jsimsa) [reply tensorflow/python/data/experimental/ops/get_single_element.py]: @kvignesh1420 this still needs to be addressed. Thanks.

Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py]: @jsimsa sorry about that. I addressed it in the commit.

**Review by jsimsa — APPROVED** [review]

**Comment by kvignesh1420** [comment]
@jsimsa the .pbtxt files have been updated!

**Review by jsimsa — APPROVED** [review]

File: tensorflow/python/data/experimental/ops/get_single_element.py [hunk 13]
```diff
@@ -20,31 +20,119 @@
 from tensorflow.python.data.ops import dataset_ops
 from tensorflow.python.data.util import structure
 from tensorflow.python.ops import gen_dataset_ops
+from tensorflow.python.util import deprecation
 from tensorflow.python.util.tf_export import tf_export
 
 
-@tf_export("data.experimental.get_single_element")
+@tf_export("data.experimental.get_single_element", "data.get_single_element")
+@deprecation.deprecated_endpoints("data.experimental.get_single_element")
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/get_single_element.py:28]: Similar to my comments on `bucket_by_sequence_length`, I would prefer exposing this in the core API as a method on the `Dataset` class. Thus the usage becomes `dataset.get_single_element()` as opposed to `tf.data.get_single_element(dataset)`. You can see how `reduce` is implemented in dataset_ops.py as it is similar to `get_single_element` in the sense that it provide a way to consume the dataset element(s) and the usage (and implementation) of `get_single_element` should ideally be consistent with `reduce`.

Comment (kvignesh1420) [reply tensorflow/python/data/experimental/ops/get_single_element.py:28]: @jsimsa sure will take a look.

**Comment by kvignesh1420** [comment]
@jsimsa the `get_single_element` API is now a part of `tf.data.Dataset`. The docstring has been modified as per the latest and deprecated APIs. Please take a look.

**Comment by jsimsa** [comment]
Could we also move `tensorflow/python/data/experimental/kernel_tests/get_single_element_test.py` to `tensorflow/python/data/kernel_tests/get_single_element_test.py` and update it so that it uses the non-experimental API as opposed to the now deprecated experimental API? Thanks.

**Comment by kvignesh1420** [comment]
@jsimsa done!

**Review by jsimsa — APPROVED** [review]

**Comment by jsimsa** [comment]
One last change. Please update https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md adding a "tf.data" subsection under the "Bug Fixes and Other Changes" section of TF 2.6.0 (next to the existing "TF Core" section) and add the following item there: "promoting `tf.data.experimental.get_single_element` API to `tf.data.Dataset.get_single_element` and deprecating the experimental endpoint"

**Comment by kvignesh1420** [comment]
@jsimsa the `RELEASE.md` file has been updated. Can we merge this before https://github.com/tensorflow/tensorflow/pull/48217 so that I can rebase and make necessary changes?

**Review by jsimsa — APPROVED** [review]

**Comment by kvignesh1420** [comment]
@jsimsa any pending changes to be addressed?

**Comment by jsimsa** [comment]
no, I am waiting for your change to RELEASE.md to be propagated to the internal copy of your PR (it has not been as of this morning) ... if it is still stuck this PM, I will fix things manually

**Comment by kvignesh1420** [comment]
ok, thanks!

**Comment by gbaned** [comment]
@kvignesh1420 Can you please resolve conflicts? Thanks!

**Comment by kvignesh1420** [comment]
@gbaned done!

**Review by jsimsa — APPROVED** [review]

**Comment by gbaned** [comment]
@kvignesh1420 Can you please resolve conflicts? Thanks!

**Comment by kvignesh1420** [comment]
@jsimsa can we get this merged so that https://github.com/tensorflow/tensorflow/pull/48217 can be updated with the release notes. Please let me know.

**Comment by jsimsa** [comment]
FYI, I have manually fixed outstanding issues and merged your change, so this PR can be closed.
